File: oop/java code/Abstraction.java

abstract  class Figure {
    double dim1, dim2;
    Figure(double a, double b){ dim1 = a; dim2 = b;}
    abstract double area();
    void show(){System.out.println("Abstract");}
    }
    class Rectangle extends Figure {
    Rectangle(double a, double b) { super(a, b);}
    double area(){ return dim1*dim2;}
    void show(){
    System.out.println("Rectangle Area: "+area());
    }
    }
    class Triangle extends Figure {
    Triangle(double a, double b) {super(a, b);}
    double area(){ return 0.5*dim1*dim2;}
    void show(){
    System.out.println("Triangle Area: "+area());
    }
    }


public class Abstraction {
    
    public static void main(String[] args) {
        Rectangle r = new Rectangle(10,7);
        Triangle t = new Triangle(10, 5);
        Figure figref;
        figref = r;
        figref.show();
        figref = t;
        figref.show();
        }
}

File: oop/c++ code/abstraction.cpp
#include <iostream>
using namespace std;
class Chudirbhai{
    int a;
    int b;
    public :
    Chudirbhai(int s, int m){
        a=s;
        b=m;
    }
};
class Figure
{
protected:
    double dim1, dim2;

public:
    Figure(double d1, double d2)
    {
        dim1 = d1;
        dim2 = d2;
    }

    virtual double area() = 0; // pure virtual function
   virtual void show (){
        cout<<"msin"<<endl;
    }
};
class Rectangle : public Figure
{
public:
    Rectangle(double d1, double d2) : Figure(d1, d2) {}

    double area()
    {
        return dim1 * dim2;
    }
    void show (){
        cout<<"rect"<<endl;
    }
};
class Triangle : public Figure
{
public:
    Triangle(double d1, double d2) : Figure(d1, d2) {}

    double area()
    {
        return dim1 * dim2 / 2;
    }
    void show (){
        cout<<"trig"<<endl;
    }
};

int main()
{
    Figure *p;
    Rectangle r(10, 7);
    Triangle t(10, 5);
    Chudirbhai x(3,2);
    Chudirbhai y =3,2;
    p = &r;
    cout << "Rectangle Area: " << p->area() << endl;
    p->show();

    p = &t;
    cout << "Triangle Area: " << p->area() << endl;
    p->show();
}
    
